core/src/main/scala/kafka/server/MetadataCache.scala

```diff
@@ -47,55 +47,83 @@ private[server] class MetadataCache(brokerId: Int) extends Logging {
-                    Errors.forException(e).code)
-              }
-          }
-          topicResponses += new TopicMetadata(topic, partitionMetadata.toSeq)
+  private def getPartitionMetadata(topic: String, protocol: SecurityProtocol): List[MetadataResponse.PartitionMetadata] = {
+    cache(topic).map { case (partitionId, partitionState) =>
+      val topicPartition = TopicAndPartition(topic, partitionId)
+
+      val leaderAndIsr = partitionState.leaderIsrAndControllerEpoch.leaderAndIsr
+      val leader = aliveBrokers.get(leaderAndIsr.leader).map(_.getNode(protocol))
+
+      val replicas = partitionState.allReplicas
+      val replicaInfo = getAliveEndpoints(replicas, protocol)
+
+      if (leader.isEmpty) {
+        debug("Error while fetching metadata for %s: leader not available".format(topicPartition))
+        new MetadataResponse.PartitionMetadata(Errors.LEADER_NOT_AVAILABLE, partitionId, Node.noNode(),
+          replicaInfo.asJava, java.util.Collections.emptyList())
+      } else {
+        val isr = leaderAndIsr.isr
+        val isrInfo = getAliveEndpoints(isr, protocol)
+
+        if (replicaInfo.size < replicas.size) {
+          debug("Error while fetching metadata for %s: replica information not available for following brokers %s"
+            .format(topicPartition, replicas.filterNot(replicaInfo.map(_.id).contains(_)).mkString(",")))
+
+          new MetadataResponse.PartitionMetadata(Errors.REPLICA_NOT_AVAILABLE, partitionId, leader.get,
+            replicaInfo.asJava, isrInfo.asJava)
+        } else if (isrInfo.size < isr.size) {
+          debug("Error while fetching metadata for %s: in sync replica information not available for following brokers %s"
+            .format(topicPartition, isr.filterNot(isrInfo.map(_.id).contains(_)).mkString(",")))
+          new MetadataResponse.PartitionMetadata(Errors.REPLICA_NOT_AVAILABLE, partitionId, leader.get,
+            replicaInfo.asJava, isrInfo.asJava)
+        } else {
+          new MetadataResponse.PartitionMetadata(Errors.NONE, partitionId, leader.get, replicaInfo.asJava,
+            isrInfo.asJava)
         }
       }
+    }.toList
+  }
+
+  def getTopicMetadata(topics: Set[String], protocol: SecurityProtocol): mutable.Buffer[MetadataResponse.TopicMetadata] = {
+    inReadLock(partitionMetadataLock) {
+      val isAllTopics = topics.isEmpty
+      val topicsRequested = if (isAllTopics) cache.keySet else getExistingTopics(topics)
+      topicsRequested.map { topic =>
+        val partitionMetadata = getPartitionMetadata(topic, protocol)
+        new MetadataResponse.TopicMetadata(Errors.NONE, topic, partitionMetadata.asJava)
+      }.toBuffer
+    }
+  }
+
+  def hasTopicMetadata(topic: String): Boolean = {
+    inReadLock(partitionMetadataLock) {
+      cache.contains(topic)
+    }
+  }
+
+  def getAllTopics(): Set[String] = {
+    inReadLock(partitionMetadataLock) {
+      Set(cache.keys.toSeq: _*)
```

Comment:
We convert the keys iterable to a `Seq` before we convert it again to a `Set`. We could reduce the number of copies by using `cache.keySet.toSet` (one copy from mutable Set to immutable Set).